### ISSUE TITLE
float not divisible by int [unsupported-operation]

### ISSUE THREAD

**Issue by tonnymadsen** [astral-sh/ty#3621]
### Summary

I have the following code, where I get an unexpected "unsupported-operator" error (on the last line). 

I would expect ty to check the operator for all the variants in the type alias.

```python
type FloatOrInt = float | int

def convert_data[ValueType: FloatOrInt](data: ValueType) -> ValueType:
    return data / 1
```

### Version

ty 0.0.41

**Comment by sharkdp** [comment]
Thank you for reporting this.

This is not related to type aliases, but rather just to bounded typevars:
```py
def convert_data[ValueType: float](data: ValueType):
   data / 1  # ty: unsupported-operator
```


By the way, note that a typevar with an upper bound of `int | float` is not the same as a constrained typevar that can either specialize to `int` or `float`. Given your issue title, it seems like the latter is maybe what you intended to use? But no matter which of these variants you really want, that code is not going to type check. `data / 1` will be of type `float` in both cases, and that is not assignable to `ValueType` (which could specialize to a subclass of `int` or `float`). So something is broken here.

**Comment by tonnymadsen** [comment]
Thanks for the speedy response.

In my original slightly more complicated case, the type alias was

```python
type FloatOrArray = float | npt.NDArray[np.float64]
```

We have a lot of functions that should work equally well on floats and Numpy arrays (and xarrays..). In all cases, the function should return the same type as given as an argument.

I can see, I can get the intended results if I use @overload, but I would really prefer not to :-)

**Comment by carljm** [comment]
ty does behave differently from other type checkers here. I think we should ignore the return-type error; all type checkers agree on that. But ty is the only one to give `unsupported-operation` here (I've removed the type aliase, since it doesn't change the behavior, and ty considers a `float` annotation the same as `float | int` anyway):

```py
def convert_data[T: float](data: T):
    data / 1  # only ty errors here
```

The ty error goes away if we just use a `float` rather than a bounded typevar:

```py
def convert_data(data: float):
    data / 1  # ty is ok with this
```

It also goes away if we use `JustFloat` (which excludes `int`):

```py
from ty_extensions import JustFloat

def convert_data[T: JustFloat](data: T):
    data / 1  # ty is ok with this
```

Based on that, I'm pretty sure this is another manifestation of our problems with calling methods on instances of typevars with a union bound. This is #2585.

Thanks for the report!
